src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Thread.NativeAot.cs

```diff
@@ -322,7 +322,23 @@ public bool Join(int millisecondsTimeout)
         /// </summary>
         internal const int OptimalMaxSpinWaitsPerSpinIteration = 64;
 
-        public static void SpinWait(int iterations) => RuntimeImports.RhSpinWait(iterations);
+        public static void SpinWait(int iterations)
+        {
+            // Max iterations to be done in one call to RhSpinWait.
+            // RhSpinWait does not switch GC modes and we want to avoid native spinning in coop mode for too long.
+            const int spinWaitBatch = 10000;
+
+            while (iterations > spinWaitBatch)
+            {
+                RuntimeImports.RhSpinWait(spinWaitBatch);
+                iterations -= spinWaitBatch;
+            }
+
+            if (iterations > 0)
```

Comment:
CoreCLR does not spin if number of iterations <= 0 . 
Same logic is used here.

https://github.com/dotnet/runtime/blob/d365ff2e866ff2b9684262d139e79ba1c4641f76/src/coreclr/vm/comsynchronizable.cpp#L1064

Reply:
Would it be better to make this:
```
if (iterations > 10000)
{
    LongSpinWait(iterations);
}
else
{
    RuntimeImports.RhSpinWait(spinWaitBatch);
}

[NoInlineable]
static void LongSpinWait(int iterations)
{
    calls RuntimeImports.RhSpinWait with full GC transition
}
```

It should make the method inlinable and reduce the fixed overhead of the spin waiting.

Reply:
I was not sure how to do `calls RuntimeImports.RhSpinWait with full GC transition` 

Can we have two PInvoke methods pointing to the same native implementation, only one with SuppressGCTransition?

Reply:
> Can we have two PInvoke methods pointing to the same native implementation, only one with SuppressGCTransition?

Yup:

https://github.com/dotnet/runtime/blob/8294512d367099942b8c7ba6e1f51a671405faf3/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/RuntimeImports.cs#L385-L389

You could add e.g.:

```csharp
        // Call RhSpinWait with a GC transition
        [LibraryImport(RuntimeLibrary, EntryPoint = "RhSpinWait")]
        [UnmanagedCallConv(CallConvs = new Type[] { typeof(CallConvCdecl) })]
        internal static partial void RhLongSpinWait(int iterations);
```

Reply:
Yes, that works just fine.